src/plugins/discord/fetcher.js

```diff
@@ -0,0 +1,147 @@
+
+export type MemberResponse = {|
+  +members: $ReadOnlyArray<Model.GuildMember>,
+  ...pageInfo,
+|};
+
+export type ReactionResponse = {|
+  +reactions: $ReadOnlyArray<Model.Reaction>,
+  ...pageInfo,
+|};
+
+export type MessageResponse = {|
+  +messages: $ReadOnlyArray<Model.Message>,
+  ...pageInfo,
+|};
+
+/**
+ * Convenience function to find the endCursor value in a response
+ * for Channels, Members, and Reactions.
+ *
+ */
+function getLastId(items: $ReadOnlyArray<any>): endCursor {
+  if (items.length > 0) {
+    return sortBy(items, (x) => x.id)[items.length - 1].id;
+  } else {
+    return "";
+  }
+}
+
+/**
+ * Fetches correct endpoint and returns re-formatted request data
+ * and pagination info. Options provide request limit info and Guild
+ * Id.
+ *
+ */
+export class Fetcher {
+  +_fetch: fetchEndpoint;
+  +_options: FetchOptions;
+
+  constructor(fetchEndpoint: fetchEndpoint, options: FetchOptions) {
+    this._fetch = fetchEndpoint;
+    this._options = options;
+  }
+
+  async guilds(): Promise<$ReadOnlyArray<Model.Guild>> {
+    const response = await this._fetch("/users/@me/guilds");
+    return response.map((x) => ({
+      id: x.id,
+      name: x.name,
+      permissions: x.permissions,
+    }));
+  }
+
+  async guild(): Promise<any> {
+    const {guildId} = this._options;
+    return this._fetch(`guilds/${guildId}`);
+  }
+
+  async channels(): Promise<$ReadOnlyArray<Model.Channel>> {
+    const {guildId} = this._options;
+    const response = await this._fetch(`/guilds/${guildId}/channels`);
+    return response.map((x) => ({
+      id: x.id,
+      name: x.name,
+      type: Model.channelTypeFromId(x.type),
+    }));
+  }
+
+  async members(after: endCursor): Promise<MemberResponse> {
+    const {membersLimit, guildId} = this._options;
+    const endpoint = `/guilds/${guildId}/members?after=${after}&limit=${membersLimit}`;
+    const response = await this._fetch(endpoint);
+    const members = response.map((x) => ({
+      user: {
+        id: x.user.id,
+        username: x.user.username,
+        discriminator: x.user.discriminator,
+        bot: x.user.bot || x.user.system || false,
+      },
+      nick: x.nick || null,
+      roles: x.roles,
+    }));
+    const hasNextPage = members.length === membersLimit;
+    const endCursor = getLastId(response.map((x) => x.user));
+    return {members, endCursor, hasNextPage};
+  }
+
+  async messages(
+    channel: Model.Snowflake,
+    after: endCursor
+  ): Promise<MessageResponse> {
+    const {messagesLimit} = this._options;
+    const endpoint = `/channels/${channel}/messages?after=${after}&limit=${messagesLimit}`;
+    const response = await this._fetch(endpoint);
+    const messages = response.map((x) => ({
+      id: x.id,
+      channelId: channel,
+      authorId: x.author.id,
+      timestampMs: Date.parse(x.timestamp),
+      content: x.content,
+      reactionEmoji: (x.reactions || []).map((r) => r.emoji),
+      nonUserAuthor: x.webhook_id != null || false,
```

Comment:
Nit: for easier reading / maintainability this could use it's own named function.

```js
isAuthoredByNonUser(rawMessage: {+webhook_id?: Snowflake}): boolean
```

Reply:
Updated